### PR TITLE
Migration to Emotion 11

### PR DESCRIPTION
**Pull request by Andarist**
Issue:

## How to test

- [x] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: lib/theming/src/index.ts [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
+import * as emotionReact from '@emotion/react';
+import type { CreateStyled, PropsOf, AddOptionalTo } from './emotion10types';
 
 import { Theme } from './types';
 
-export type { StyledComponent } from '@emotion/styled';
-export { Global, keyframes, css, jsx, ClassNames } from '@emotion/core';
-export type { CSSObject, Keyframes } from '@emotion/core';
+export type { StyledComponent } from './emotion10types';
+
```
Comment (ndelangen) [inline lib/theming/src/index.ts:11]: We need a comment here so we remove this eventually when we release 7.0

**Comment by ndelangen** [comment]
amazing work, what's left to do?

**Comment by Andarist** [comment]
I got to the bottom of those changes in the baselines.

TLDR - there was probably a parsing bug in Stylis v3 (or in `@emotion/stylis`) and the previous version was incorrect. The new one is actually correct, both on the technical level and also visually.

Those styles [here](https://github.com/storybookjs/storybook/blob/729a87023b3fc9a1fdaa67e5794f24cc0d2513f0/lib/components/src/blocks/ArgsTable/ArgsTable.tsx#L19-L189) were producing this css~ string:
```sass
&&{border-collapse:collapse;border-spacing:0;color:#FFFFFF;td, th{padding:0;border:none;vertical-align:top;text-overflow:ellipsis;}font-size:13px;line-height:20px;text-align:left;width:100%;margin-top:25px;margin-bottom:40px;thead th:first-of-type, td:first-of-type{width:25%;}th:first-of-type, td:first-of-type{padding-left:20px;}th:nth-of-type(2), td:nth-of-type(2){width:35%;}td:nth-of-type(3){width:15%;}th:last-of-type, td:last-of-type{padding-right:20px;width:25%;}th{color:rgba(255,255,255,0.55);padding-top:10px;padding-bottom:10px;padding-left:15px;padding-right:15px;}td{padding-top:10px;padding-bottom:10px;&:not(:first-of-type){padding-left:15px;padding-right:15px;}&:last-of-type{padding-right:20px;}}margin-left:1px;margin-right:1px;tr:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{td:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-top-left-radius:4px;}td:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-top-right-radius:4px;}}tr:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{td:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-bottom-left-radius:4px;}td:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */, th:last-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */{border-bottom-right-radius:4px;}}tbody{box-shadow:rgba(0, 0, 0, 0.20) 0 2px 5px 1px,
          rgba(255,255,255,0.15) 0 0 0 1px;border-radius:4px;@media not all and (min-resolution:.001dpcm){@supports (-webkit-appearance:none){border-width:1px;border-style:solid;border-color:rgba(255,255,255,0.15);}}tr{background:transparent;overflow:hidden;&:not(:first-child/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */){border-top-width:1px;border-top-style:solid;border-top-color:#404040;}}td{background:#333333;}}}th span, td span, td button{display:inline;background-color:rgba(255,255,255,.1);animation:animation-r0iffl 1.5s ease-in-out infinite;color:transparent;box-shadow:none;border-radius:0;}
```
In here basically, the whole block is wrapped with `&& { ... }`, so everything inside should produce rules with a "double scope". However, with Emotion 10 we got such a rule out of this:
```css
.css-1eaai6p td:last-child,.css-1eaai6p th:last-child{border-top-right-radius:4px;}
```
We can notice that it doesn't contain the double scope which is definitely not correct. From the very same string the new version produces such a rule:
```css
.css-1eaai6p.css-1eaai6p tr:last-child td:last-child,.css-1eaai6p.css-1eaai6p tr:last-child th:last-child{border-bottom-right-radius:4px;}
```
We can also notice that the old one has only a single `:last-child` whereas the correct one has `:last-child :last-child` in it.

This caused a minor visual bug in the old version and that was fixed with this dependency upgrade.

A funny thing is that if I either:
- prettify this CSS string
- OR remove "ignoreSsrWarning" comments

then the whole thing parses correctly with the old version. So there was definitely some edge-case bug in the parsing there.

**Comment by ndelangen** [comment]
I think this replaces this PR: https://github.com/storybookjs/storybook/pull/13300

File: examples/react-ts/src/__snapshots__/storyshots.test.ts.snap [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
 exports[`Storyshots Demo/AccountForm Standard 1`] = `
-.emotion-15 {
+.emotion-0 {
```
Comment (Andarist) [inline examples/react-ts/src/__snapshots__/storyshots.test.ts.snap:4]: those snapshots just have different generated "indices" for those masked class names

File: lib/components/src/controls/Object.tsx [hunk 3]
```diff
@@ -244,7 +244,7 @@ const getCustomStyleFunction: (theme: Theme) => JsonTreeProps['getStyle'] = (the
 });
 
 export const ObjectControl: FC<ObjectProps> = ({ name, value, onChange }) => {
-  const theme = useTheme<Theme>();
```
Comment (Andarist) [inline lib/components/src/controls/Object.tsx:247]: This shows a small, subtle, breaking change. I'm open to reverting this. 

Previously you had to provide a type param here. Now you don't have to - so it can be just skipped. However, this means that the previously valid code is no longer valid:
```ts
const theme = useTheme<Theme>();
```
We could adjust the signature of `useTheme` to look something like that:
```ts
declare function useTheme<TTheme = Theme>(): TTheme;
```
This way you wouldn't have to provide a generic, but also you could provide it. While you shouldn't be parametrizing it with just "any" Theme object... this would avoid that small breaking change.

File: lib/theming/src/emotion10types.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
+import { Interpolation } from '@emotion/react';
```
Comment (Andarist) [inline lib/theming/src/emotion10types.ts:1]: it's basically just a copy-paste (adjusted one) from `@emotion/core@10`

File: lib/components/src/controls/Object.tsx (hunk 3 above)
Comment (ndelangen) [reply lib/components/src/controls/Object.tsx:247]:
```ts
declare function useTheme<TTheme = Theme>(): TTheme;
```
Sounds like a good idea!

**Review by shilman — COMMENTED** [review]
So exciting!! A couple bits

File: lib/cli/src/versions.ts [hunk 5]
```diff
@@ -1,59 +1,59 @@
...
-  "@storybook/store": "6.5.0-alpha.52",
-  "@storybook/svelte": "6.5.0-alpha.52",
-  "@storybook/theming": "6.5.0-alpha.52",
-  "@storybook/ui": "6.5.0-alpha.52",
-  "@storybook/vue": "6.5.0-alpha.52",
-  "@storybook/vue3": "6.5.0-alpha.52",
-  "@storybook/web-components": "6.5.0-alpha.52"
-}
\ No newline at end of file
+  '@storybook/addon-a11y': '6.5.0-alpha.52',
```
[the remark this answers is not in the document]
Comment (Andarist) [reply lib/cli/src/versions.ts:3]: right, this change was caused by your precommit hook that got executed when I was committing merge conflict resolution

File: lib/components/src/controls/Object.tsx (hunk 3 above)
Comment (Andarist) [reply lib/components/src/controls/Object.tsx:247]: I've added this change, the whole PR should be ready to merge now :)

File: lib/theming/src/emotion10types.ts (hunk 4 above)
Comment (ndelangen) [reply lib/theming/src/emotion10types.ts:1]: I added a little not telling us to remove this when releasing 7.0

File: lib/theming/src/index.ts [hunk 6]
```diff
@@ -1,23 +1,32 @@
...
-import { CacheProvider } from '@emotion/core';
-import emotionStyled, { CreateStyled } from '@emotion/styled';
+import emotionStyled from '@emotion/styled';
+import * as emotionReact from '@emotion/react';
+// TODO: when releasing 7.0 this should be removed and code should be upgraded to use Emotion 11 types
+import type { CreateStyled, PropsOf, AddOptionalTo } from './emotion10types';
 
 import { Theme } from './types';
 
-export { CacheProvider };
```
Comment (ndelangen) [inline lib/theming/src/index.ts:9]: @Andarist I think this CacheProvider export is now removed, can we add this back?

**Review by ndelangen — CHANGES_REQUESTED** [review]

File: lib/theming/src/index.ts (hunk 6 above)
Comment (ndelangen) [reply lib/theming/src/index.ts:9]: actually, never mind it IS exported on line 13!

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
I'll merge as soon as CI is green(-ish)

**Comment by ndelangen** [comment]
going to ignore this error:
> [vite]: Rollup failed to resolve import "react-dom/client" from "node_modules/@storybook/react/dist/esm/client/preview/render.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`

for not, since @IanVS is fixing that issue (caused by https://github.com/storybookjs/storybook/pull/17215)
